design/ceph/ceph-orchestrator-osd-add-remove.md

```diff
@@ -0,0 +1,266 @@
+     `ceph-volume` should report success without taking any actions
+2. Drive Groups - support a blacklist of devices
+   1. Should be able to blacklist individual devices by the device path
+   2. The blacklist may be able to support other ways of specifying devices to blacklist, but above
+     is the minimum requirement
+
+### Adding OSDs
+This design proposes to add a section to the `CephCluster` resource's `storage` configuration
+section to allow disks to be added via a Drive Groups specification. The spec is proposed below with
+behavior described to follow.
+
+```yaml
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+spec:
+  storage: # preexisting
+    driveGroups: # proposed new
+      - spec: < string, JSON or YAML blob to pass to ceph-volume for OSD creation >
+```
+
+Each individual `driveGroups` item's `spec` will be a simple string containing a JSON blob or YAML
+doc string for the Drive Group. `ceph-volume` feature request #1 will be used here to allow Rook to
+pass along the Drive Group JSON/YAML string without modification or inspection to `ceph-volume` when
+running the OSD provisioning job. OSD creation via Drive Groups is given priority and executed
+before any other CRD-based OSD creation methods are used (the other CRD-based creations methods are
+still used after the Drive Group method).
+
+Drive Groups should be defined only at the cluster level, not at node level. Drive Groups support a
+glob matching pattern for nodes, so every drive group should be issued on every node. Rook will not
+suspect an error for non-matching nodes due to `ceph-volume` feature 1-3.
+
+Users of `CephCluster` via Kubernetes Manifest will now be able to supply their own Drive Groups
+spec to Rook if they so desire to allow themselves greater flexibility in OSD management. Rook will
+treat this user spec as a declarative intent for the cluster.
+
+The Rook Orchestrator Mgr Module should be able to report the Drive Groups in the Rook configuration
+to the user, and the Orchestrator Module will have the ability to add new drive groups, remove
+existing drive groups, and modify existing drive groups. The user should be advised to only manage
+drive groups via one method: either via the `CephCluster` resource themselves manually or via the
+Orchestrator CLI or Dashboard interface.
+
+#### Reporting Status
+Rook should report failure messages back to the Rook Orchestrator Mgr Module via a `CephCluster`
+`status` mechanism so that the Orchestrator Module can take appropriate actions ot notify the user
+of failure. Rook may report successful statuses back if needed by the Orchestrator Module; however,
+limiting the information reported back will keep the `CephCluster` resource's `status` field smaller
+and less unwieldy for users.
+
+```yaml
+# first draft suggestions for status mechanism
+apiVersion: ceph.rook.io/v1
+kind: CephCluster
+status:
+  driveGroupOSDCreations: # proposed new
```

Comment:
That sounds like a good idea to me. I'll work on adding in the basic skeleton for such status in this design.

Reply:
In my view errors creating OSD's (by any method) is something that can be only logged because basically is a sequence of operations to reach one target (create the OSD), and we can have problem in any part of the sequence. To have a good log file indicating what is happening in each moment is for sure the best way to know what is the problem. 
Other uncomfortable thing is that if we having a problem not solved in one OSD creation can result in hundred of error messages stored in the cluster CRD.

Reply:
I disagree. I think it could be important to have an at-a-glance view of failures. Even if the failure message isn't super helpful, the message should have enough info for the user to know where to look for more info. As long as the message includes which OSD.ID is failing, if it was during preparation or runtime, and on which node, the OSD or OSD prepare pod logs are the obvious next target for inspection, and those will have the full logs.

Reply:
Added to appendix a  in the doc